manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,740 @@
+    write to that memory location in the execution trace, then "x" precedes "y"
+    in happens-before order. For atomic locations, "compare_and_set",
+    "fetch_and_add", "exchange", "incr" and "decr" are considered to perform
+    both a read and a write.
+  \item If "x" is "Domain.spawn f" and "y" is the first action in the newly
+    spawned domain executing "f", then "x" precedes "y" in happens-before
+    order.
+  \item If "x" is the last action in a domain "d" and "y" is "Domain.join
+    d", then "x" precedes "y" in happens-before order.
+  \item If "x" is an unlock operation on a mutex, and "y" is any subsequent
+    operation on the mutex in the execution trace, then "x" precedes "y" in
+    happens-before order.
+\end{itemize}
+
+\subsection{s:datarace}{Data race}
+
+In a given trace, two actions are said to be \emph{conflicting} if they access
+the same non-atomic location, and at least one is a write.
+
+We say that a program has a \emph{data race} if there exists some execution
+trace of the program with two conflicting actions and there does not exist a
+happens-before relationship between the conflicting accesses. A program without
+data races is said to be \emph{correctly synchronised}.
+
+\subsection{ss:drf_sc}{DRF-SC}
+
+\textbf{DRF-SC guarantee:} A program without data races will only exhibit
+sequentially consistent behaviours.
+
+DRF-SC is a strong guarantee for the programmers. Programmers can use
+sequential reasoning (executing one inter-domain action to identify whether
+their program has a data race. In particular, they do not need to reason about
+reorderings described in section~\ref{s:why_relaxed_memory} in order to determine
+whether their program has a data race. Once the determination that a particular
+program is data race free is made, they do not need to worry about reorderings
+in their code.
+
+For programs with data races, OCaml memory model ensures that there are no
+out-of-thin-air (OOTA) values. Reads always return a value written to a memory
+location (including possibly be the initial write). Data races do not cause
+programs to crash.
+
+\section{s:drf_reasoning}{Reasoning with DRF-SC}
+
+In this section, we will look at examples of using DRF-SC for program
+reasoning. In this section, we will use the functions with names "dN" to
+represent domains executing in parallel with other domains.
+
+Here is a simple example with a data race:
+
+\begin{caml_example*}{verbatim}
+(* Has data race *)
+let r = ref 0
+let d1 () = ref 1
+let d2 () = !r
+\end{caml_example*}
+
+"r" is a non-atomic reference. The two domains race to access the reference,
+and "d1" is a write. Since there is no happens-before relationship between the
+conflicting accesses, there is a data race.
+
+Both of the programs that we had seen in the section~\ref{s:why_relaxed_memory}
+have have data races, and hence, exhibit non sequentially consistent
+behaviours.
```

Comment:
Hum, data racy programs do not necessarily have non-SC behaviours, as illustrated by the above simple example. Tentative suggestion:
```suggestion
have have data races. It is no surprise that they exhibit non sequentially consistent
behaviours.
```

Reply:
Thanks.